File: libraries/java/src/main/java/com/smcp/security/SMCPSecurityFramework.java
package com.smcp.security;

import com.smcp.security.auth.JWTAuthenticator;
import com.smcp.security.auth.MFAManager;
import com.smcp.security.auth.RBACManager;
import com.smcp.security.audit.AuditLogger;
import com.smcp.security.crypto.SMCPCrypto;
import com.smcp.security.exception.*;
import com.smcp.security.model.*;
import com.smcp.security.ratelimit.AdaptiveRateLimiter;
import com.smcp.security.validation.InputValidator;
import com.smcp.security.ai.AIImmuneSystem;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Instant;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Main SMCP Security Framework that orchestrates all security components.
 * 
 * This class provides a comprehensive security framework for Model Context Protocol
 * implementations, including input validation, authentication, authorization,
 * rate limiting, encryption, AI-based threat detection, and audit logging.
 * 
 * @author SMCP Security Team
 * @version 1.0.0
 * @since 1.0.0
 */
public class SMCPSecurityFramework {
    
    private static final Logger logger = LoggerFactory.getLogger(SMCPSecurityFramework.class);
    
    private final SecurityConfig config;
    private final InputValidator inputValidator;
    private final JWTAuthenticator jwtAuthenticator;
    private final MFAManager mfaManager;
    private final RBACManager rbacManager;
    private final AdaptiveRateLimiter rateLimiter;
    private final SMCPCrypto crypto;
    private final AuditLogger auditLogger;
    private final AIImmuneSystem aiImmune;
    private final SecurityMetrics metrics;
    
    // Metrics tracking
    private final AtomicLong totalRequests = new AtomicLong(0);
    private final AtomicLong blockedRequests = new AtomicLong(0);
    private final AtomicLong threatsDetected = new AtomicLong(0);
    private final AtomicLong authenticationFailures = new AtomicLong(0);
    private final AtomicLong rateLimitHits = new AtomicLong(0);
    private volatile double averageResponseTime = 0.0;
    private volatile Instant lastThreatDetected;
    
    /**
     * Creates a new SMCP Security Framework with default configuration.
     */
    public SMCPSecurityFramework() {
        this(SecurityConfig.defaultConfig());
    }
    
    /**
     * Creates a new SMCP Security Framework with the specified configuration.
     * 
     * @param config the security configuration
     * @throws SecurityConfigurationException if the configuration is invalid
     */
    public SMCPSecurityFramework(SecurityConfig config) {
        this.config = validateConfig(config);
        this.metrics = new SecurityMetrics();
        
        // Initialize components based on configuration
        this.inputValidator = config.isEnableInputValidation() ? 
            new InputValidator(config.getValidationStrictness()) : null;
            
        this.jwtAuthenticator = new JWTAuthenticator(
            config.getJwtSecret(),
            config.getJwtExpirySeconds()
        );
        
        this.mfaManager = config.isEnableMFA() ? new MFAManager() : null;
        
        this.rbacManager = config.isEnableRBAC() ? new RBACManager() : null;
        if (this.rbacManager != null) {
            setupDefaultRoles();
        }
        
        this.rateLimiter = config.isEnableRateLimiting() ? 
            new AdaptiveRateLimiter(config.getDefaultRateLimit(), config.isAdaptiveLimits()) : null;
            
        this.crypto = config.isEnableEncryption() ? new SMCPCrypto() : null;
        
        this.auditLogger = config.isEnableAuditLogging() ? 
            new AuditLogger(config.getLogLevel()) : null;
            
        this.aiImmune = config.isEnableAIImmune() ? 
            new AIImmuneSystem(config.getAnomalyThreshold(), config.isLearningMode()) : null;
        
        logger.info("SMCP Security Framework initialized with configuration: {}", config);
    }
    
    /**
     * Validates an MCP request against all configured security policies.
     * 
     * @param request the MCP request to validate
     * @return the validated (and potentially sanitized) request
     * @throws ValidationException if the request fails validation
     * @throws SecurityException if a security threat is detected
     */
    public MCPRequest validateRequest(MCPRequest request) throws SecurityException {
        long startTime = System.currentTimeMillis();
        totalRequests.incrementAndGet();
        
        try {
            logger.debug("Validating MCP request: {}", request.getId());
            
            // Input validation
            MCPRequest validatedRequest = request;
            if (inputValidator != null) {
                ValidationResult result = inputValidator.validate(request);
                
                if (!result.isValid()) {
                    blockedRequests.incrementAndGet();
                    
                    if (auditLogger != null) {
                        auditLogger.logValidationFailure(request, result.getErrors());
                    }
                    
                    throw new ValidationException("Request validation failed", result.getErrors());
                }
                
                validatedRequest = result.getSanitizedRequest();
            }
            
            // AI immune system check
            if (aiImmune != null) {
                ThreatAnalysis analysis = aiImmune.analyzeRequest(validatedRequest);
                
                if (analysis.getRiskScore() > config.getAnomalyThreshold()) {
                    threatsDetected.incrementAndGet();
                    blockedRequests.incrementAndGet();
                    lastThreatDetected = Instant.now();
                    
                    if (auditLogger != null) {
                        auditLogger.logThreatDetection(validatedRequest, analysis);
                    }
                    
                    throw new ThreatDetectedException(
                        "High-risk request detected: " + analysis.getThreatType(),
                        analysis.getRiskScore()
                    );
                }
            }
            
            // Update metrics
            updateAverageResponseTime(System.currentTimeMillis() - startTime);
            
            logger.debug("Request validation completed successfully: {}", request.getId());
            return validatedRequest;
            
        } catch (SecurityException e) {
            if (auditLogger != null) {
                auditLogger.logSecurityEvent(request, e);
            }
            throw e;
        }
    }
    
    /**
     * Authenticates user credentials and returns a JWT token.
     * 
     * @param credentials the user credentials
     * @return JWT token for authenticated user
     * @throws AuthenticationException if authentication fails
     */
    public String authenticateUser(AuthCredentials credentials) throws AuthenticationException {
        try {
            logger.debug("Authenticating user: {}", credentials.getUsername());
            
            // Basic credential validation
            if (credentials.getUsername() == null || credentials.getUsername().trim().isEmpty() ||
                credentials.getPassword() == null || credentials.getPassword().trim().isEmpty()) {
                authenticationFailures.incrementAndGet();
                throw new AuthenticationException("Username and password are required");
            }
            
            // Verify credentials (in real implementation, check against database)
            if (!verifyCredentials(credentials)) {
                authenticationFailures.incrementAndGet();
                
                if (auditLogger != null) {
                    auditLogger.logAuthenticationFailure(credentials.getUsername(), "Invalid credentials");
                }
                
                throw new AuthenticationException("Invalid credentials");
            }
            
            // MFA verification if enabled
            if (mfaManager != null && credentials.getMfaCode() != null) {
                boolean mfaValid = mfaManager.verifyToken(credentials.getUsername(), credentials.getMfaCode());
                
                if (!mfaValid) {
                    authenticationFailures.incrementAndGet();
                    
                    if (auditLogger != null) {
                        auditLogger.logAuthenticationFailure(credentials.getUsername(), "Invalid MFA code");
                    }
                    
                    throw new AuthenticationException("Invalid MFA code");
                }
            } else if (mfaManager != null) {
                throw new AuthenticationException("MFA code required");
            }
            
            // Generate JWT token
            UserContext userContext = getUserContext(credentials.getUsername());
            String token = jwtAuthenticator.generateToken(userContext);
            
            if (auditLogger != null) {
                auditLogger.logAuthenticationSuccess(credentials.getUsername(), credentials.getMfaCode() != null);
            }
            
            logger.debug("User authenticated successfully: {}", credentials.getUsername());
            return token;
            
        } catch (AuthenticationException e) {
            logger.warn("Authentication failed for user: {}", credentials.getUsername());
            throw e;
        }
    }
    
    /**
     * Checks if a user is authorized to perform a specific action.
     * 
     * @param userId the user ID
     * @param action the action to authorize
     * @return true if authorized, false otherwise
     * @throws AuthorizationException if authorization check fails
     */
    public boolean authorizeAction(String userId, String action) throws AuthorizationException {
        if (rbacManager == null) {
            return true; // Authorization disabled
        }
        
        try {
            boolean hasPermission = rbacManager.hasPermission(userId, action);
            
            if (auditLogger != null) {
                auditLogger.logAuthorizationAttempt(userId, action, hasPermission);
            }
            
            return hasPermission;
            
        } catch (Exception e) {
            logger.error("Authorization check failed for user {} and action {}", userId, action, e);
            throw new AuthorizationException("Authorization check failed", action);
        }
    }
    
    /**
     * Processes a secure request with full security checks.
     * 
     * @param request the MCP request
     * @param userId the user ID
     * @param sessionToken the JWT session token
     * @return the processed result
     * @throws SecurityException if any security check fails
     */
    public Object processSecureRequest(MCPRequest request, String userId, String sessionToken) 
            throws SecurityException {
        long startTime = System.currentTimeMillis();
        
        try {
            logger.debug("Processing secure request: {} for user: {}", request.getId(), userId);
            
            // Validate JWT token
            UserContext userContext = jwtAuthenticator.verifyToken(sessionToken);
            
            if (!userContext.getUserId().equals(userId)) {
                throw new AuthenticationException("Token user mismatch");
            }
            
            // Rate limiting check
            if (rateLimiter != null) {
                if (!rateLimiter.isAllowed(userId)) {
                    rateLimitHits.incrementAndGet();
                    
                    if (auditLogger != null) {
                        auditLogger.logRateLimitExceeded(userId, request);
                    }
                    
                    throw new RateLimitException("Rate limit exceeded for user: " + userId);
                }
            }
            
            // Validate the request
            MCPRequest validatedRequest = validateRequest(request);
            
            // Authorization check
            String action = request.getMethod();
            if (!authorizeAction(userId, action)) {
                throw new AuthorizationException("Access denied for action: " + action, action);
            }
            
            // Process the request (placeholder implementation)
            Object result = processRequest(validatedRequest, userContext);
            
            // Log successful request
            if (auditLogger != null) {
                auditLogger.logRequestSuccess(userId, validatedRequest, result, 
                    System.currentTimeMillis() - startTime);
            }
            
            logger.debug("Secure request processed successfully: {}", request.getId());
            return result;
            
        } catch (SecurityException e) {
            if (auditLogger != null) {
                auditLogger.logRequestFailure(userId, request, e, 
                    System.currentTimeMillis() - startTime);
            }
            throw e;
        }
    }
    
    /**
     * Gets current security metrics.
     * 
     * @return security metrics
     */
    public SecurityMetrics getSecurityMetrics() {
        return SecurityMetrics.builder()
            .totalRequests(totalRequests.get())
            .blockedRequests(blockedRequests.get())
            .threatsDetected(threatsDetected.get())
            .averageResponseTime(averageResponseTime)
            .activeUsers(0L) // Would be calculated from active sessions
            .rateLimitHits(rateLimitHits.get())
            .authenticationFailures(authenticationFailures.get())
            .lastThreatDetected(lastThreatDetected)
            .build();
    }
    
    // Private helper methods
    
    private SecurityConfig validateConfig(SecurityConfig config) {
        if (config == null) {
            throw new SecurityConfigurationException("Security configuration cannot be null");
        }
        
        if (config.getDefaultRateLimit() <= 0) {
            throw new SecurityConfigurationException("Rate limit must be positive");
        }
        
        if (config.getAnomalyThreshold() < 0.0 || config.getAnomalyThreshold() > 1.0) {
            throw new SecurityConfigurationException("Anomaly threshold must be between 0.0 and 1.0");
        }
        
        if (config.getJwtExpirySeconds() <= 0) {
            throw new SecurityConfigurationException("JWT expiry must be positive");
        }
        
        return config;
    }
    
    private void setupDefaultRoles() {
        if (rbacManager != null) {
            rbacManager.addRole("admin", java.util.List.of(
                "read", "write", "delete", "admin",
                "tools:*", "resources:*", "prompts:*"
            ));
            
            rbacManager.addRole("user", java.util.List.of(
                "read", "tools:list", "tools:call",
                "resources:list", "resources:read",
                "prompts:list", "prompts:get"
            ));
            
            rbacManager.addRole("readonly", java.util.List.of(
                "read", "tools:list", "resources:list", "prompts:list"
            ));
        }
    }
    
    private boolean verifyCredentials(AuthCredentials credentials) {
        // In a real implementation, this would check against a database
        // For demo purposes, accept any non-empty credentials
        return credentials.getUsername() != null && !credentials.getUsername().trim().isEmpty() &&
               credentials.getPassword() != null && !credentials.getPassword().trim().isEmpty();
    }
    
    private UserContext getUserContext(String username) {
        // In a real implementation, this would fetch from database
        return UserContext.builder()
            .userId(username)
            .username(username)
            .role("user")
            .permissions(config.getDefaultPermissions())
            .sessionId(java.util.UUID.randomUUID().toString())
            .lastActivity(Instant.now())
            .build();
    }
    
    private Object processRequest(MCPRequest request, UserContext userContext) {
        // This is a placeholder - in a real implementation, this would
        // delegate to the actual MCP server implementation
        return MCPResponse.builder()
            .jsonrpc("2.0")
            .id(request.getId())
            .result(java.util.Map.of(
                "message", "Request processed successfully",
                "method", request.getMethod(),
                "user", userContext.getUserId()
            ))
            .build();
    }
    
    private void updateAverageResponseTime(long newTime) {
        long totalReqs = totalRequests.get();
        if (totalReqs > 0) {
            averageResponseTime = ((averageResponseTime * (totalReqs - 1)) + newTime) / totalReqs;
        }
    }
}

File: libraries/csharp/SMCPSecurityFramework.cs
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;
using SMCP.Security.Authentication;
using SMCP.Security.Authorization;
using SMCP.Security.Cryptography;
using SMCP.Security.Exceptions;
using SMCP.Security.Models;
using SMCP.Security.RateLimit;
using SMCP.Security.Validation;
using SMCP.Security.Audit;
using SMCP.Security.AI;
using System.Collections.Concurrent;

namespace SMCP.Security;

/// <summary>
/// Main SMCP Security Framework that orchestrates all security components.
/// Provides comprehensive security for Model Context Protocol implementations
/// including input validation, authentication, authorization, rate limiting,
/// encryption, AI-based threat detection, and audit logging.
/// </summary>
public class SMCPSecurityFramework : ISMCPSecurityFramework
{
    private readonly SMCPSecurityOptions _options;
    private readonly ILogger<SMCPSecurityFramework> _logger;
    private readonly IInputValidator? _inputValidator;
    private readonly IJWTAuthenticator _jwtAuthenticator;
    private readonly IMFAManager? _mfaManager;
    private readonly IRBACManager? _rbacManager;
    private readonly IAdaptiveRateLimiter? _rateLimiter;
    private readonly ISMCPCrypto? _crypto;
    private readonly IAuditLogger? _auditLogger;
    private readonly IAIImmuneSystem? _aiImmune;
    
    // Metrics tracking
    private long _totalRequests = 0;
    private long _blockedRequests = 0;
    private long _threatsDetected = 0;
    private long _authenticationFailures = 0;
    private long _rateLimitHits = 0;
    private double _averageResponseTime = 0.0;
    private DateTime? _lastThreatDetected;
    private readonly object _metricsLock = new();
    
    /// <summary>
    /// Initializes a new instance of the SMCPSecurityFramework.
    /// </summary>
    /// <param name="options">Security configuration options</param>
    /// <param name="logger">Logger instance</param>
    /// <param name="inputValidator">Input validator (optional)</param>
    /// <param name="jwtAuthenticator">JWT authenticator</param>
    /// <param name="mfaManager">MFA manager (optional)</param>
    /// <param name="rbacManager">RBAC manager (optional)</param>
    /// <param name="rateLimiter">Rate limiter (optional)</param>
    /// <param name="crypto">Cryptography service (optional)</param>
    /// <param name="auditLogger">Audit logger (optional)</param>
    /// <param name="aiImmune">AI immune system (optional)</param>
    public SMCPSecurityFramework(
        IOptions<SMCPSecurityOptions> options,
        ILogger<SMCPSecurityFramework> logger,
        IInputValidator? inputValidator = null,
        IJWTAuthenticator? jwtAuthenticator = null,
        IMFAManager? mfaManager = null,
        IRBACManager? rbacManager = null,
        IAdaptiveRateLimiter? rateLimiter = null,
        ISMCPCrypto? crypto = null,
        IAuditLogger? auditLogger = null,
        IAIImmuneSystem? aiImmune = null)
    {
        _options = options.Value ?? throw new ArgumentNullException(nameof(options));
        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
        
        ValidateConfiguration(_options);
        
        // Initialize components based on configuration
        _inputValidator = _options.EnableInputValidation ? inputValidator : null;
        _jwtAuthenticator = jwtAuthenticator ?? throw new ArgumentNullException(nameof(jwtAuthenticator));
        _mfaManager = _options.EnableMFA ? mfaManager : null;
        _rbacManager = _options.EnableRBAC ? rbacManager : null;
        _rateLimiter = _options.EnableRateLimiting ? rateLimiter : null;
        _crypto = _options.EnableEncryption ? crypto : null;
        _auditLogger = _options.EnableAuditLogging ? auditLogger : null;
        _aiImmune = _options.EnableAIImmune ? aiImmune : null;
        
        // Setup default roles if RBAC is enabled
        if (_rbacManager != null)
        {
            SetupDefaultRoles();
        }
        
        _logger.LogInformation("SMCP Security Framework initialized with configuration: {@Options}", _options);
    }
    
    /// <summary>
    /// Parameterless constructor for basic usage.
    /// </summary>
    public SMCPSecurityFramework() : this(
        Options.Create(new SMCPSecurityOptions()),
        new LoggerFactory().CreateLogger<SMCPSecurityFramework>())
    {
    }
    
    /// <summary>
    /// Validates an MCP request against all configured security policies.
    /// </summary>
    /// <param name="request">The MCP request to validate</param>
    /// <param name="cancellationToken">Cancellation token</param>
    /// <returns>The validated (and potentially sanitized) request</returns>
    /// <exception cref="ValidationException">Thrown when request validation fails</exception>
    /// <exception cref="ThreatDetectedException">Thrown when a security threat is detected</exception>
    public async Task<MCPRequest> ValidateRequestAsync(MCPRequest request, CancellationToken cancellationToken = default)
    {
        var startTime = DateTime.UtcNow;
        Interlocked.Increment(ref _totalRequests);
        
        try
        {
            _logger.LogDebug("Validating MCP request: {RequestId}", request.Id);
            
            // Input validation
            var validatedRequest = request;
            if (_inputValidator != null)
            {
                var validationResult = await _inputValidator.ValidateAsync(request, cancellationToken);
                
                if (!validationResult.IsValid)
                {
                    Interlocked.Increment(ref _blockedRequests);
                    
                    if (_auditLogger != null)
                    {
                        await _auditLogger.LogValidationFailureAsync(request, validationResult.Errors, cancellationToken);
                    }
                    
                    throw new ValidationException("Request validation failed", validationResult.Errors);
                }
                
                validatedRequest = validationResult.SanitizedRequest ?? request;
            }
            
            // AI immune system check
            if (_aiImmune != null)
            {
                var analysis = await _aiImmune.AnalyzeRequestAsync(validatedRequest, cancellationToken);
                
                if (analysis.RiskScore > _options.AnomalyThreshold)
                {
                    Interlocked.Increment(ref _threatsDetected);
                    Interlocked.Increment(ref _blockedRequests);
                    _lastThreatDetected = DateTime.UtcNow;
                    
                    if (_auditLogger != null)
                    {
                        await _auditLogger.LogThreatDetectionAsync(validatedRequest, analysis, cancellationToken);
                    }
                    
                    throw new ThreatDetectedException(
                        $"High-risk request detected: {analysis.ThreatType}",
                        analysis.RiskScore
                    );
                }
            }
            
            // Update metrics
            var processingTime = (DateTime.UtcNow - startTime).TotalMilliseconds;
            UpdateAverageResponseTime(processingTime);
            
            _logger.LogDebug("Request validation completed successfully: {RequestId}", request.Id);
            return validatedRequest;
        }
        catch (SecurityException ex)
        {
            if (_auditLogger != null)
            {
                await _auditLogger.LogSecurityEventAsync(request, ex, cancellationToken);
            }
            throw;
        }
    }
    
    /// <summary>
    /// Authenticates user credentials and returns a JWT token.
    /// </summary>
    /// <param name="credentials">User credentials</param>
    /// <param name="cancellationToken">Cancellation token</param>
    /// <returns>JWT token for authenticated user</returns>
    /// <exception cref="AuthenticationException">Thrown when authentication fails</exception>
    public async Task<string> AuthenticateUserAsync(AuthCredentials credentials, CancellationToken cancellationToken = default)
    {
        try
        {
            _logger.LogDebug("Authenticating user: {Username}", credentials.Username);
            
            // Basic credential validation
            if (string.IsNullOrWhiteSpace(credentials.Username) || string.IsNullOrWhiteSpace(credentials.Password))
            {
                Interlocked.Increment(ref _authenticationFailures);
                throw new AuthenticationException("Username and password are required");
            }
            
            // Verify credentials (in real implementation, check against database)
            if (!await VerifyCredentialsAsync(credentials, cancellationToken))
            {
                Interlocked.Increment(ref _authenticationFailures);
                
                if (_auditLogger != null)
                {
                    await _auditLogger.LogAuthenticationFailureAsync(credentials.Username, "Invalid credentials", cancellationToken);
                }
                
                throw new AuthenticationException("Invalid credentials");
            }
            
            // MFA verification if enabled
            if (_mfaManager != null && !string.IsNullOrEmpty(credentials.MfaCode))
            {
                var mfaValid = await _mfaManager.VerifyTokenAsync(credentials.Username, credentials.MfaCode, cancellationToken);
                
                if (!mfaValid)
                {
                    Interlocked.Increment(ref _authenticationFailures);
                    
                    if (_auditLogger != null)
                    {
                        await _auditLogger.LogAuthenticationFailureAsync(credentials.Username, "Invalid MFA code", cancellationToken);
                    }
                    
                    throw new AuthenticationException("Invalid MFA code");
                }
            }
            else if (_mfaManager != null)
            {
                throw new AuthenticationException("MFA code required");
            }
            
            // Generate JWT token
            var userContext = await GetUserContextAsync(credentials.Username, cancellationToken);
            var token = await _jwtAuthenticator.GenerateTokenAsync(userContext, cancellationToken);
            
            if (_auditLogger != null)
            {
                await _auditLogger.LogAuthenticationSuccessAsync(credentials.Username, !string.IsNullOrEmpty(credentials.MfaCode), cancellationToken);
            }
            
            _logger.LogDebug("User authenticated successfully: {Username}", credentials.Username);
            return token;
        }
        catch (AuthenticationException)
        {
            _logger.LogWarning("Authentication failed for user: {Username}", credentials.Username);
            throw;
        }
    }
    
    /// <summary>
    /// Checks if a user is authorized to perform a specific action.
    /// </summary>
    /// <param name="userId">User ID</param>
    /// <param name="action">Action to authorize</param>
    /// <param name="cancellationToken">Cancellation token</param>
    /// <returns>True if authorized, false otherwise</returns>
    /// <exception cref="AuthorizationException">Thrown when authorization check fails</exception>
    public async Task<bool> AuthorizeActionAsync(string userId, string action, CancellationToken cancellationToken = default)
    {
        if (_rbacManager == null)
        {
            return true; // Authorization disabled
        }
        
        try
        {
            var hasPermission = await _rbacManager.HasPermissionAsync(userId, action, cancellationToken);
            
            if (_auditLogger != null)
            {
                await _auditLogger.LogAuthorizationAttemptAsync(userId, action, hasPermission, cancellationToken);
            }
            
            return hasPermission;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Authorization check failed for user {UserId} and action {Action}", userId, action);
            throw new AuthorizationException("Authorization check failed", action);
        }
    }
    
    /// <summary>
    /// Processes a secure request with full security checks.
    /// </summary>
    /// <param name="request">MCP request</param>
    /// <param name="userId">User ID</param>
    /// <param name="sessionToken">JWT session token</param>
    /// <param name="cancellationToken">Cancellation token</param>
    /// <returns>Processed result</returns>
    /// <exception cref="SecurityException">Thrown when any security check fails</exception>
    public async Task<object> ProcessSecureRequestAsync(MCPRequest request, string userId, string sessionToken, CancellationToken cancellationToken = default)
    {
        var startTime = DateTime.UtcNow;
        
        try
        {
            _logger.LogDebug("Processing secure request: {RequestId} for user: {UserId}", request.Id, userId);
            
            // Validate JWT token
            var userContext = await _jwtAuthenticator.VerifyTokenAsync(sessionToken, cancellationToken);
            
            if (userContext.UserId != userId)
            {
                throw new AuthenticationException("Token user mismatch");
            }
            
            // Rate limiting check
            if (_rateLimiter != null)
            {
                if (!await _rateLimiter.IsAllowedAsync(userId, cancellationToken))
                {
                    Interlocked.Increment(ref _rateLimitHits);
                    
                    if (_auditLogger != null)
                    {
                        await _auditLogger.LogRateLimitExceededAsync(userId, request, cancellationToken);
                    }
                    
                    throw new RateLimitException($"Rate limit exceeded for user: {userId}");
                }
            }
            
            // Validate the request
            var validatedRequest = await ValidateRequestAsync(request, cancellationToken);
            
            // Authorization check
            var action = request.Method;
            if (!await AuthorizeActionAsync(userId, action, cancellationToken))
            {
                throw new AuthorizationException($"Access denied for action: {action}", action);
            }
            
            // Process the request (placeholder implementation)
            var result = await ProcessRequestAsync(validatedRequest, userContext, cancellationToken);
            
            // Log successful request
            if (_auditLogger != null)
            {
                var processingTime = (DateTime.UtcNow - startTime).TotalMilliseconds;
                await _auditLogger.LogRequestSuccessAsync(userId, validatedRequest, result, processingTime, cancellationToken);
            }
            
            _logger.LogDebug("Secure request processed successfully: {RequestId}", request.Id);
            return result;
        }
        catch (SecurityException ex)
        {
            if (_auditLogger != null)
            {
                var processingTime = (DateTime.UtcNow - startTime).TotalMilliseconds;
                await _auditLogger.LogRequestFailureAsync(userId, request, ex, processingTime, cancellationToken);
            }
            throw;
        }
    }
    
    /// <summary>
    /// Gets current security metrics.
    /// </summary>
    /// <returns>Security metrics</returns>
    public SecurityMetrics GetSecurityMetrics()
    {
        lock (_metricsLock)
        {
            return new SecurityMetrics
            {
                TotalRequests = _totalRequests,
                BlockedRequests = _blockedRequests,
                ThreatsDetected = _threatsDetected,
                AverageResponseTime = _averageResponseTime,
                ActiveUsers = 0, // Would be calculated from active sessions
                RateLimitHits = _rateLimitHits,
                AuthenticationFailures = _authenticationFailures,
                LastThreatDetected = _lastThreatDetected
            };
        }
    }
    
    // Private helper methods
    
    private static void ValidateConfiguration(SMCPSecurityOptions options)
    {
        if (options.DefaultRateLimit <= 0)
        {
            throw new SecurityConfigurationException("Rate limit must be positive");
        }
        
        if (options.AnomalyThreshold < 0.0 || options.AnomalyThreshold > 1.0)
        {
            throw new SecurityConfigurationException("Anomaly threshold must be between 0.0 and 1.0");
        }
        
        if (options.JwtExpirySeconds <= 0)
        {
            throw new SecurityConfigurationException("JWT expiry must be positive");
        }
    }
    
    private void SetupDefaultRoles()
    {
        if (_rbacManager == null) return;
        
        _rbacManager.AddRole("admin", new[]
        {
            "read", "write", "delete", "admin",
            "tools:*", "resources:*", "prompts:*"
        });
        
        _rbacManager.AddRole("user", new[]
        {
            "read", "tools:list", "tools:call",
            "resources:list", "resources:read",
            "prompts:list", "prompts:get"
        });
        
        _rbacManager.AddRole("readonly", new[]
        {
            "read", "tools:list", "resources:list", "prompts:list"
        });
    }
    
    private async Task<bool> VerifyCredentialsAsync(AuthCredentials credentials, CancellationToken cancellationToken)
    {
        // In a real implementation, this would check against a database
        // For demo purposes, accept any non-empty credentials
        await Task.Delay(10, cancellationToken); // Simulate async database call
        return !string.IsNullOrWhiteSpace(credentials.Username) && !string.IsNullOrWhiteSpace(credentials.Password);
    }
    
    private async Task<UserContext> GetUserContextAsync(string username, CancellationToken cancellationToken)
    {
        // In a real implementation, this would fetch from database
        await Task.Delay(10, cancellationToken); // Simulate async database call
        
        return new UserContext
        {
            UserId = username,
            Username = username,
            Role = "user",
            Permissions = _options.DefaultPermissions.ToList(),
            SessionId = Guid.NewGuid().ToString(),
            LastActivity = DateTime.UtcNow
        };
    }
    
    private async Task<object> ProcessRequestAsync(MCPRequest request, UserContext userContext, CancellationToken cancellationToken)
    {
        // This is a placeholder - in a real implementation, this would
        // delegate to the actual MCP server implementation
        await Task.Delay(10, cancellationToken); // Simulate processing
        
        return new MCPResponse
        {
            JsonRpc = "2.0",
            Id = request.Id,
            Result = new Dictionary<string, object>
            {
                ["message"] = "Request processed successfully",
                ["method"] = request.Method,
                ["user"] = userContext.UserId
            }
        };
    }
    
    private void UpdateAverageResponseTime(double newTime)
    {
        lock (_metricsLock)
        {
            var totalReqs = _totalRequests;
            if (totalReqs > 0)
            {
                _averageResponseTime = ((_averageResponseTime * (totalReqs - 1)) + newTime) / totalReqs;
            }
        }
    }
}
